Support dynamic list of allowed ActionCable connection origins

multiplegeorges opened the pull request:
### Summary

With this change, ActionCable config now support assigning a proc to config.allowed_request_origins. This allows the use of dynamically defined lists of allowed origins.

```ruby
config.allowed_request_origins = -> { Sites.pluck(:domain) }
```

### Gotchas

If an origin is *removed* from the list, existing connections will remain open until a server restart.

It is up to the developer to add an origin identifier to their connections and then use the RemoteConnections class to find and disconnect them.

This isn't ideal. Any suggestions on how to handle this in a better and secure-by-default way are very welcome.

mrhead committed:
Do not enable disabled elements for XHR redirects

Fixes #29473.

fatkodima committed:
Fix `transaction` reverting for migrations

[fatkodima & David Verhasselt]

Commit:
If association is a hash-like object preloading fails

If you pass a hash-like object to preload associations (for example ActionController::Parameters)
preloader will fail with the ArgumentError.
This change allows passing objects that may be converted to a Hash or String into a preloader

yahonda committed:
Add `Style/RedundantFreeze` to remove redudant `.freeze`

Since Rails 6.0 will support Ruby 2.4.1 or higher
`# frozen_string_literal: true` magic comment is enough to make string object frozen.
This magic comment is enabled by `Style/FrozenStringLiteralComment` cop.

* Exclude these files not to auto correct false positive `Regexp#freeze`
 - 'actionpack/lib/action_dispatch/journey/router/utils.rb'
 - 'activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb'

It has been fixed by https://github.com/rubocop-hq/rubocop/pull/6333
Once the newer version of RuboCop released and available at Code Climate these exclude entries should be removed.

* Replace `String#freeze` with `String#-@` manually if explicit frozen string objects are required

 - 'actionpack/test/controller/test_case_test.rb'
 - 'activemodel/test/cases/type/string_test.rb'
 - 'activesupport/lib/active_support/core_ext/string/strip.rb'
 - 'activesupport/test/core_ext/string_ext_test.rb'
 - 'railties/test/generators/actions_test.rb'

gmcgibbon committed:
Add deprecation warning when String#first and String#last receive negative integers

[Gannon McGibbon + Eric Turner]

multiplegeorges commented:
The reported failure seems to be unrelated to any changes contained in this PR.

Jcambass committed:
respect path_only option when an array is passed into url_for

The url_for method is now extracting the path_only option in order to determine if polymorphic_path or polymorphic_url should be called.

If the path_only option is not set it will be set to true unless the host option is set. This behaviour is the same as when a Hash or Params object is passed.

To support this unifying the code responsible for setting this default value has been extracted into a private method

al2o3cr committed:
failing test for eager loading

cherry-picked from https://github.com/rails/rails/pull/33763

duduribeiro committed:
[ci skip] Recommend testing information displayed over object stored

Since `assigns` is not placed by default with Rails, we suggested to
test if the right information was displayed instead of suggesting to
test if the right object was stored in the template.

kamipo committed:
Don't expose internal methods in the associations

kamipo committed:
Refactor counter cache to extract `decrement_counters_before_last_save` on the belongs_to association

yskkin committed:
Deprecate ActionDispatch::Http::ParameterFilter in favor of ActiveSupport::ParameterFilter

faucct committed:
ActiveRecord::Associations::Preloader should not fail to preload through missing records

sakshi-keepworks committed:
Remove private def

kaspth committed:
Merge pull request #33949 from sjain1107/no-private-def

Remove private def

kaspth committed:
Merge pull request #33829 from mtsmfm/encode-filename

Encode Content-Disposition filenames on send_data and send_file

kamipo committed:
`Persistence#increment!` requires an attribute argument which is incremented

kaspth committed:
Extract `instrument` method.

Similar to Action View's and Action Controller's instrument helpers.

eugeneius committed:
Eagerly define attribute methods in production

The attribute methods for a model are currently defined lazily the first
time that model is instantiated, even when `config.eager_load` is true.
This means the first request to use each model incurs the cost, which
usually involves a database round trip to fetch the schema definition.

By defining the attribute methods for all models while the application
is booting, we move that work out of any individual request. When using
a forking web server, this also reduces the number of times the schema
definition is queried by doing it once in the parent process instead of
from each forked process during their first request.

utilum committed:
update mini_magick to the latest version

v 4.9.2 avoids some warnings that show in our tests.

Although Gemfile.lock currently specifies 4.8.0, the CI already runs 4.9.2
so 4.8.0 warnings do not show in our CI logs.

kamipo committed:
Merge pull request #33960 from utilum/update_minimagick_version_to_4.9.2

update mini_magick to the latest version

pskl committed:
Improve phrasing of one sentence in the contributing guide

kamipo committed:
Merge pull request #33959 from eugeneius/eager_load_attribute_methods

Eagerly define attribute methods in production

gmcgibbon committed:
Stringify database configurations

palkan committed:
Add ActionCable::Channel::TestCase

ActionCable::Channel::TestCase provides an ability
to unit-test channel classes.

There are several reasons to write unit/functional cable tests:
- Access control (who has access to the channel? who can perform action and with which argument?
- Frontend-less applications have no system tests at all–and we still need a way to test channels logic.

See also #27191

tenderlove committed:
Eagerly build the routing helper module after routes are committed

This commit eagerly builds the route helper module after the routes have
been drawn and finalized.  This allows us to cache the helper module but
not have to worry about people accessing the module while route
definition is "in-flight", and automatically deals with cache
invalidation as the module is regenerated anytime someone redraws the
routes.

The restriction this commit introduces is that the url helper module can
only be accessed *after* the routes are done being drawn.

Refs #24554 and #32892

tenderlove committed:
Remove deprecated catch-all route in the AV tests

This commit removes a deprecated catch-all route in the AV tests.  It
defines and includes the necessary routes for each test such that we
don't need the catch-all anymore.

This also helps push us toward #33970

tenderlove committed:
Add hack to deal with warnings

We should be able to remove this once the catch-all route is gone from
AP

tenderlove committed:
make bot happy

tenderlove committed:
This patch removes deprecated catch-all routes from AM

It also removes a monkey patch from AM::Base

rafaelfranca committed:
Merge pull request #33964 from pskl/master

Improve phrasing of one sentence in the contributing guide

rafaelfranca committed:
Merge pull request #33953 from wagurano/guides_markdown_custom_header_id

Fix rails guides markdown.rb and renderer.rb to use custom header id

Commit:
Fix a content_for test description

marshall-lee committed:
Abandon TOP support.

Initially, `TOP` was introduced to support `limit` for MSSQL database.
Unlike PostgreSQL/MySQL/SQLite, MSSQL does not have native `LIMIT`/`OFFSET` support.
The commit adding `TOP` is 1a246f71616cf246a75ef6cbdb56032e43d4e643.

However, it figured out that `TOP` implementation was weak and it's not sufficient
to also support `OFFSET`, then `TOP` was substituted with
`ROW_NUMBER()` subquery in be48ed3071fd6524d0145c4ad3faeb4aafe3eda3.
This is a well known trick in MSSQL -
https://stackoverflow.com/questions/2135418/equivalent-of-limit-and-offset-for-sql-server.

So now we don't need this `visit_Arel_Nodes_Top` at all.
It does nothing useful but also adds an extra space after `SELECT` when `LIMIT` is being
used for **any** database.

lzap committed:
Added ActionController::Parameters.each_value method

kamipo committed:
Merge pull request #33977 from marshall-lee/abandon-top-support

Abandon TOP support.

tenderlove committed:
Merge pull request #33973 from rails/remove-catch-all

Remove deprecated catch-all route in the AV tests

tenderlove committed:
Merge pull request #33974 from rails/remove-catch-all-from-am

This patch removes deprecated catch-all routes from AM

rafaelfranca committed:
Change the empty block style to have space inside of the block

rafaelfranca committed:
Fix more offences

kamipo committed:
Remove `counter_cache_target` which is no longer called

`counter_cache_target` is called only when updated counter cache in
replacing target, but it was already removed at #33913.

rafaelfranca committed:
Merge pull request #33975 from JuanitoFatas/jf.fix-test-name

[CaptureHelperTest] Fix a content_for test description

tenderlove committed:
Allow helpers to be deferred until the routes have been finalized

ActiveStorage::BaseController subclasses ActionController::Base.
ActionController::Base has an "inherited" hook set that includes the
routing helpers to any subclass of AC::Base.  Since
ActiveStorage::BaseController is a subclass of AC::Base, it will get
routing helpers included automatically.  Unfortunately, when the
framework is eagerly loaded, ActiveStorage::BaseController is loaded
*before* the applications routes are loaded which means it attempts to
include an "in flight" module so it gets an exception.

This commit allows a class that's interested in being extended with
routing helpers register itself such that when the routes are finalized,
it will get the helpers included.  If the routes are already finalized,
then the helpers get included immediately.

rafaelfranca committed:
Merge pull request #33968 from gmcgibbon/stringify_db_configurations

Stringify database configurations

gmcgibbon committed:
Raise an error when loading all fixtures from nil fixture_path

[Gannon McGibbon + Max Albrecht]

y-yagi committed:
Fix "warning: shadowing outer local variable - config"

matt17r committed:
Add missing rdoc +code+ tags

yahonda committed:
Use the same option for create database statements between Raketask and travis.rb

kamipo committed:
Merge pull request #33987 from yahonda/specify_database_characterset_at_travis

Use the same option for create database statements

kamipo committed:
Update counter cache in memory if parent target is existed

Fixes #19550.

kamipo committed:
Revert "Remove `counter_cache_target` which is no longer called"

This reverts commit 376ffe0ea2e59dc51461122210729c05a10fb443.

Since 38fae1f, `association.increment_counters` is called without
inflated parent target if inverse_of is disabled.
In that case, that commit would cause extra queries to inflate parent.

kamipo committed:
Remove force parent loading when counter cache child is created/destroyed

`association.increment_counters` and `association.decrement_counters`
works regardless of parent target is loaded or not.

Related 52e11e462f6114a4d12225c639c5f501f0ffec7a.

eileencodes committed:
Merge pull request #31819 from bpohoriletz/master

If association is a hash-like object preloading fails

rafaelfranca committed:
Merge pull request #33983 from gmcgibbon/raise_if_fixture_path_blank

Raise an error when loading all fixtures from nil fixture_path

tenderlove committed:
Eager load supports individual files

Also we want to eager load routes before anything else.

rafaelfranca committed:
Merge pull request #30676 from artofhuman/import-assert-attrs-error-message

Improve error message when assign wrong attributes to model

tenderlove committed:
Routes from Engine Railties should not be an infinite loop

tenderlove committed:
Merge branch 'master' into eager-url-helpers

* master:
  Remove force parent loading when counter cache child is created/destroyed
  Raise an error when loading all fixtures from nil fixture_path
  Revert "Remove `counter_cache_target` which is no longer called"
  Update counter cache in memory if parent target is existed
  If association is a hash-like object preloading fails
  Use the same option for create database statements between Raketask and travis.rb
  Fix "warning: shadowing outer local variable - config"
  Remove `counter_cache_target` which is no longer called
  Fix more offences
  Change the empty block style to have space inside of the block
  Fix a content_for test description
  Stringify database configurations
  Improve error message when assign wrong attributes to model

gmcgibbon committed:
Add migrations_paths option to model generator

tenderlove committed:
Merge pull request #33970 from rails/eager-url-helpers

Eagerly build the routing helper module after routes are committed

Edouard-chin committed:
Add a way to check for subset of arguments when performing jobs:

- When calling `assert_performed_with`/`assert_enqueued_with`, the
  +args+ needs to match exactly what the job get passed.

  Some jobs can have lot of arguments, or even a simple hash argument
  has many key. This is not convenient to test as most tests doesn't
  need to check if the arguments matches perfectly.

  This PR make it possible to only check if a subset of arguments were
  passed to the job.

KarlAmort committed:
Removed invalid -X flag for pg_dump

dark-panda committed:
Use -X when loading structure.sql via psql

kamipo committed:
Merge pull request #33997 from MatthiasWinkelmann/pg_dump

Removed invalid -X flag for pg_dump, fixes #33996

kamipo committed:
Merge pull request #34001 from dark-panda/ignore-psqlrc-when-using-psql

Use -X when loading structure.sql via psql

kamipo committed:
Avoid extra touch queries when counter cache is updated

Since counter cache handles touch option too.

gmcgibbon committed:
Fix optionally scoped root route unscoped access

palkan committed:
Remove Rails 5.0 workaround from ActionCable::Channel::TestCase

The hack was merged from action-cable-testing gem by mistake.
We don't need it in Rails 6.

(cherry picked from commit 92030ec4b4309835ed0e792229984a1f0a044cef)

rafaelfranca committed:
Merge pull request #33994 from gmcgibbon/rails_g_model_migrations_paths

Add migrations_paths option to model generator

eileencodes committed:
Merge pull request #31604 from fatkodima/reverting-transaction

Fix `transaction` reverting for migrations

rafaelfranca committed:
Merge pull request #33979 from lzap/master

Added ActionController::Parameters.each_value methods

kamipo committed:
Make `update_all` preparable

Before:

```
  Pet Update All (0.8ms)  UPDATE `pets` LEFT OUTER JOIN `toys` ON `toys`.`pet_id` = `pets`.`pet_id` SET `pets`.`name` = 'Bob' WHERE `toys`.`name` = ?  [["name", "Bone"]]
```

After:

```
  Pet Update All (1.1ms)  UPDATE `pets` LEFT OUTER JOIN `toys` ON `toys`.`pet_id` = `pets`.`pet_id` SET `pets`.`name` = ? WHERE `toys`.`name` = ?  [["name", "Bob"], ["name", "Bone"]]
```

gmcgibbon committed:
Merge pull request #33986 from matt17r/patch-1

Add missing rdoc +code+ tags [ci skip]

rafaelfranca committed:
Merge pull request #34003 from palkan/fix/cable-test-case-rails5-hack

Remove Rails 5.0 workaround from ActionCable::Channel::TestCase

simonc committed:
Fixing an edge case when using objects as constraints

This PR fixes an issue when the following situation occurs.
If you define a class like this

    class MyConstraint
      def call(*args)
        # for some reason this is defined
      end

      def matches?(*args)
        # checking the args
      end
    end

and try to use it as a constraint

    get "/", to: "home#show", constraints: MyConstraint.new

if its `matches?` method returns `false` there will be an error for the
mapper will ask for the constraint arity, thinking it is a proc, lambda
or method.

This PR checks for the presence of the `arity` method on the constraint
calling it only if present, preventing the error while keeping the basic
behavior.

rafaelfranca committed:
Merge pull request #33995 from Edouard-chin/ec-active-job-subset-arg

Add a way to check for subset of arguments when performing jobs:

kamipo committed:
Make `update_counters` preparable

Before:

```
  Topic Update All (0.4ms)  UPDATE `topics` SET `topics`.`replies_count` = COALESCE(`topics`.`replies_count`, 0) + 1, `topics`.`updated_at` = '2018-09-27 18:34:05.068774' WHERE `topics`.`id` = ?  [["id", 7]]

```

After:

```
  Topic Update All (0.4ms)  UPDATE `topics` SET `topics`.`replies_count` = COALESCE(`topics`.`replies_count`, 0) + ?, `topics`.`updated_at` = ? WHERE `topics`.`id` = ?  [["replies_count", 1], ["updated_at", 2018-09-27 18:55:05 UTC], ["id", 7]]

```

tgturner committed:
Array with single item correctly uses cache_key

kamipo committed:
Remove `visit_Fixnum` and `visit_Bignum`

Since Ruby 2.4 unified Fixnum and Bignum into Integer.

kamipo committed:
Revert "record who created the node when $DEBUG is true"

This reverts commit a1b72178714fbf0033fe076b7e51f57eff152bdd.

kamipo committed:
Extract `Arel.arel_node?` helper method

rafaelfranca committed:
Merge pull request #34004 from simonc/constraints-call-matches-edge-case

Fixing an edge case when using objects as constraints

kamipo committed:
Merge pull request #34006 from kamipo/remove_debug_code

Revert "record who created the node when $DEBUG is true"

rafaelfranca committed:
Merge pull request #34002 from gmcgibbon/fix_deeply_nested_scoped_root

Fix optionally scoped root route unscoped access

gmcgibbon committed:
Add probot autolabeler config

rafaelfranca committed:
Add configuration to no-response probot plugin

[ci skip]

rafaelfranca committed:
Merge pull request #34007 from gmcgibbon/automatic_labeling

Add probot autolabeler config

jaynetics committed:
Handle more unsafe String methods (#33990)

* Handle more unsafe String methods

* Fix codeclimate issue

* Revert stylistic change

[Janosch Müller + Rafael Mendonça França]

rafaelfranca committed:
Merge pull request #31441 from mrhead/rails-ujs-xhr-redirects

Do not enable disabled elements for XHR redirects

gmcgibbon committed:
Merge pull request #33833 from duduribeiro/patch-2

[ci skip] Do not recommend testing stored object in template

kamipo committed:
Merge pull request #33880 from robbertbrak/master

Clarify transactional behavior of after_commit and after_rollback callbacks

[ci skip]

Edouard-chin committed:
LoggerSilence doesn't require concurrent:

- LoggerThreadSafeLevel does nowaday since 2518bda97cbbcb33dc9a92e70d5b01c09e64d12d

kamipo committed:
Merge pull request #34010 from Edouard-chin/ec-uneeded-require

LoggerSilence doesn't require concurrent:

abraham-chan committed:
Fix HashWithIndifferentAccess#without bug

bogdan committed:
Bugfix ActiveRecord::Relation#merge special case of from clause

When one relation is merged into another that has a different base class
merging `from_clause` causes invalid SQL to be generated

kamipo committed:
Merge pull request #34013 from bogdan/relation-merge-from

Bugfix ActiveRecord::Relation#merge special case of from clause

kamipo committed:
Merge pull request #33570 from mohitnatoo/queue_adapter_name_doc

Documentation for queue_adapter_name method in ActiveJob::QueueAdapter

[ci skip]

kamipo committed:
Merge pull request #33348 from ruralocity/update-validation-contexts-guide

Update guide for validation custom contexts [ci skip]

utilum committed:
Update backburner to the latest version

version 1.5.0 includes to warning fixes that show on tests:
https://github.com/nesquena/backburner/commit/cd9670146bca4df83ae2aacb95b3e6d1c1c1eee8
https://github.com/nesquena/backburner/commit/4dfa22bda1685ca1afa9f7d474df471ca6364e69

rafaelfranca committed:
Merge pull request #34017 from utilum/update_backburner_to_latest_version

Update backburner to the latest version

rafaelfranca committed:
Merge pull request #34012 from abraham-chan/indifferent-access-without

Fix HashWithIndifferentAccess#without bug

mohitnatoo committed:
Documentation clarity in ActiveJob::TestHelper [ci skip] (#33571)

* Documentation clarity in ActiveJob::TestHelper [ci skip]

* Documentation for options

[Mohit Natoo + Rafael Mendonça França]

tgturner committed:
Updates expires_in docs

Previously the documentation for expires_in suggested that the option
was available for all cache stores. This clarifies that the behavior can
be supported, but is not always supported.

gmcgibbon committed:
Refactor migrations_path command option to database

eileencodes committed:
Merge pull request #34021 from gmcgibbon/change_migrations_paths_command_option_to_db

Refactor migrations_path command option to database

eileencodes committed:
Basic API for connection switching

This PR adds the ability to 1) connect to multiple databases in a model,
and 2) switch between those connections using a block.

To connect a model to a set of databases for writing and reading use
the following API. This API supercedes `establish_connection`. The
`writing` and `reading` keys represent handler / role names and
`animals` and `animals_replica` represents the database key to look up
the configuration hash from.

```
class AnimalsBase < ApplicationRecord
  connects_to database: { writing: :animals, reading: :animals_replica }
end
```

Inside the application - outside the model declaration - we can switch
connections with a block call to `connected_to`.

If we want to connect to a db that isn't default (ie readonly_slow) we
can connect like this:

Outside the model we may want to connect to a new database (one that is
not in the default writing/reading set) - for example a slow replica for
making slow queries. To do this we have the `connected_to` method that
takes a `database` hash that matches the signature of `connects_to`. The
`connected_to` method also takes a block.

```
AcitveRecord::Base.connected_to(database: { slow_readonly: :primary_replica_slow }) do
  ModelInPrimary.do_something_thats_slow
end
```

For models that are already loaded and connections that are already
connected, `connected_to` doesn't need to pass in a `database` because
you may want to run queries against multiple databases using a specific
role/handler.

In this case `connected_to` can take a `role` and use that to swap on
the connection passed. This simplies queries - and matches how we do it
in GitHub. Once you're connected to the database you don't need to
re-connect, we assume the connection is in the pool and simply pass the
handler we'd like to swap on.

```
ActiveRecord::Base.connected_to(role: :reading) do
  Dog.read_something_from_dog
  ModelInPrimary.do_something_from_model_in_primary
end
```

y-yagi committed:
Merge pull request #34020 from tgturner/update-cache-docs

Updates expires_in docs

y-yagi committed:
Reset `ActiveRecord::Base.configurations` to the value before

y-yagi committed:
Make `test_initialize_with_invalid_attribute` work correctly

Originally specified attributes were only normal values, and
`ActiveRecord::MultiparameterAssignmentErrors` did not occur.

In addition, an assertion is performed only on rescue, even if an
exception does not occur, the test passes. To avoid this use `assert_raise`.

kamipo committed:
Remove `visit_Fixnum` and `visit_Bignum`

Follow up ae406cd633dab2cafbc0d1bb5922d1ca40056ea0.

kamipo committed:
Use private attr_reader in Arel

No longer needed workaround for Ruby 2.2 "private attribute?" warning.

Related 6d63b5e49a399fe246afcebad45c3c962de268fa.

kamipo committed:
`visitor.accept` doesn't handle `&block`

kamipo committed:
`SQLString#compile` is no longer used since 53521a9e39b9d8af4165d7703c36dc905f1f8f67

kamipo committed:
Handle DELETE with LIMIT in Arel

MySQL supports DELETE with LIMIT and ORDER BY.

https://dev.mysql.com/doc/refman/8.0/en/delete.html

Before:

```
  Post Destroy (1.0ms)  DELETE FROM `posts` WHERE `posts`.`id` IN (SELECT `id` FROM (SELECT `posts`.`id` FROM `posts` WHERE `posts`.`author_id` = ? ORDER BY `posts`.`id` ASC LIMIT ?) __active_record_temp)  [["author_id", 1], ["LIMIT", 1]]
```

After:

```
  Post Destroy (0.4ms)  DELETE FROM `posts` WHERE `posts`.`author_id` = ? ORDER BY `posts`.`id` ASC LIMIT ?  [["author_id", 1], ["LIMIT", 1]]
```

kamipo committed:
Handle UPDATE/DELETE with OFFSET in Arel

kamipo committed:
Place `PartialQuery` and `PartialQueryCollector` in the same file

lsylvester committed:
Require the except hash extension before trying to alias it

y-yagi committed:
Merge pull request #34031 from lsylvester/require-except-before-aliasing-it

Require the except hash extension before trying to alias it

lsylvester committed:
make actionview templates marshalable so that they can be serialized during the parallel tests (#34030)

Commit:
Explain why we have explicit marshaling methods

kamipo committed:
Merge pull request #32031 from yahonda/remove_redundant_freeze

Add `Style/RedundantFreeze` to remove redudant `.freeze`

kamipo committed:
Merge pull request #33256 from ilkkao/ilkkao/remove-unused-params-option

Don't handle params option in a special way in url_for helper

kamipo committed:
Merge pull request #23593 from meinac/add_index_option_for_change_table

index option added for change_table migrations

Commit:
Make Webpacker the default JavaScript compiler for Rails 6 (#33079)

* Use Webpacker by default on new apps

* Stop including coffee-rails by default

* Drop using a js_compressor by default

* Drop extra test for coffeescript inclusion by default

* Stick with skip_javascript to signify skipping webpack

* Don't install a JS runtime by default any more

* app/javascript will be the new default directory for JS

* Make it clear that this is just for configuring the default Webpack framework setup now

* Start using the Webpack tag in the default layout

* Irrelevant test

* jQuery is long gone

* Stop having asset pipeline compile default application.js

* Add rails-ujs by default to the Webpack setup

* Add Active Storage JavaScript to application.js pack by default

* Consistent quoting

* Add Turbolinks to default pack

* Add Action Cable to default pack

Need some work on how to set the global consumer that channels will
work with. @javan?

* Require all channels by default and use a separate consumer stub

* Channel generator now targets Webpack style

* Update task docs to match new generator style

* Use uniform import style

* Drop the JS assets generator

It was barely helpful as it was. It’s no longer helpful in a Webpacked
world. Sayonara!

* Add app/javascript to the stats directories

* Simpler import style

Which match the other imports.

* Address test failures from dropping JS compilation (and compression)

* webpacker-default: Modify `AssetsGeneratorTest`

Before:

```
$ bin/test test/generators/assets_generator_test.rb
Run options: --seed 46201

F

Failure:
AssetsGeneratorTest#test_assets [/Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/assets_generator_test.rb:12]:
Expected file "app/assets/javascripts/posts.js" to exist, but does not

bin/test /Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/assets_generator_test.rb:10

.

Finished in 0.031343s, 63.8101 runs/s, 95.7152 assertions/s.
2 runs, 3 assertions, 1 failures, 0 errors, 0 skips
```

After:

```
$ bin/test test/generators/assets_generator_test.rb
Run options: --seed 43571

..

Finished in 0.030370s, 65.8545 runs/s, 65.8545 assertions/s.
2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
```

* webpacker-default: Modify `ChannelGeneratorTest`

Before:

```
$ bin/test test/generators/channel_generator_test.rb
Run options: --seed 8986

.F

Failure:
ChannelGeneratorTest#test_channel_with_multiple_actions_is_created [/Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:43]:
Expected file "app/assets/javascripts/channels/chat.js" to exist, but does not

bin/test /Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:34

.F

Failure:
ChannelGeneratorTest#test_channel_is_created [/Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:29]:
Expected file "app/assets/javascripts/channels/chat.js" to exist, but does not

bin/test /Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:22

E

Error:
ChannelGeneratorTest#test_cable_js_is_created_if_not_present_already:
Errno::ENOENT: No such file or directory @ apply2files - /Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/fixtures/tmp/app/assets/javascripts/cable.js

bin/test /Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:60

F

Failure:
ChannelGeneratorTest#test_channel_suffix_is_not_duplicated [/Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:87]:
Expected file "app/assets/javascripts/channels/chat.js" to exist, but does not

bin/test /Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:80

F

Failure:
ChannelGeneratorTest#test_channel_on_revoke [/Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:77]:
Expected file "app/assets/javascripts/cable.js" to exist, but does not

bin/test /Users/ttanimichi/ghq/github.com/ttanimichi/rails/railties/test/generators/channel_generator_test.rb:68

Finished in 0.064384s, 108.7227 runs/s, 481.4861 assertions/s.
7 runs, 31 assertions, 4 failures, 1 errors, 0 skips
```

After:

```
$ bin/test test/generators/channel_generator_test.rb
Run options: --seed 44857

.......

Finished in 0.060243s, 116.1961 runs/s, 697.1764 assertions/s.
7 runs, 42 assertions, 0 failures, 0 errors, 0 skips
```

* Fix shared generator tests.

* webpacker-default: Modify `ControllerGeneratorTest`

The JS assets generator was dropped. ref. https://github.com/rails/rails/commit/46215b179483d3e4d264555f5a4952f43eb8142a

* Revert "Simpler import style". It's currently failing with an error of "TypeError: undefined is not an object (evaluating '__WEBPACK_IMPORTED_MODULE_2_activestorage___default.a.start')". Waiting for @javan to have a look.

This reverts commit 5d3ebb71059f635d3756cbda4ab9752027e09256.

* require webpacker in test app

* Add webpacker without making the build hang/timeout. (#33640)

* use yarn workspaces to allow for installing unreleased packages and only generate js/bootsnap when required

* no longer need to have webpacker in env templates as webpacker moved this config to yml file

* Fix rubocop violation

* Got the test passing for the running scaffold

* update expected lines of code

* update middleware tests to account for webpacker

* disable js in plugins be default to get the tests passing (#34009)

* clear codeclimate report issues

* Anything newer than currently released is good

* Use Webpacker development version during development of Rails

* Edge should get development webpacker as well

* Add changelog entry for Webpacker change

kamipo committed:
Remove duplicated before_install entries in .travis.yml

It has incorrectly been re-added at #33079.

Related #33861.

oniofchaos commented:
If you rebase against latest master, that often cleans up gremlins in the test build.

Edouard-chin committed:
Deprecate the `LoggerSilence` constant:

- I found this weird that the LoggerSilence wasn't using the
  `ActiveSupport` namespace (AFAIK all other classes have it).

  This PR deprecate the use of `LoggerSilence` for
  `ActiveSupport::LoggerSilence` instead.

rtlechow committed:
Fix typos

Fixes some typos.

y-yagi committed:
Remove unnecessary `coffee-rails`

Since we no longer use coffeescript in the generated files by
4838c1716a0340137d858fab49bf460e23be5a4b, this is no longer necessary.

y-yagi committed:
Merge pull request #34046 from rtlechow/fix-typos

Fix typos

sharang-d committed:
Fix spellings for 'unmarshall(ing/ed)' & 'marshall(ing/ed)'

kamipo committed:
Merge pull request #34049 from sharang-d/marshalling-text-changes

Fix spellings for 'unmarshall(ing/ed)' & 'marshall(ing/ed)'

rafaelfranca committed:
Merge pull request #33058 from gmcgibbon/string_first_last_negative_deprecation

Add deprecation warning when String#first and String#last receive neg…

gmcgibbon committed:
Prefix Module#parent, Module#parents, and Module#parent_name with module

prathamesh-sonpatki committed:
Update the docs now that webpacker is default [ci skip]

prathamesh-sonpatki committed:
Turbolinks is enabled by default in new apps [ci skip]

gmcgibbon committed:
Fix call sites

rafaelfranca committed:
Merge pull request #34045 from Edouard-chin/ec-logger-silence-namespace

Deprecate the `LoggerSilence` constant:

rafaelfranca committed:
Merge pull request #34053 from prathamesh-sonpatki/update-js-docs

Update js docs after the webpacker changes

rafaelfranca committed:
Merge pull request #34051 from gmcgibbon/module_parent_method_rename

Prefix Module#parent, Module#parents, and Module#parent_name with module

kamipo committed:
Move UPDATE/DELETE with JOIN handling to the Arel side

Edouard-chin committed:
Fix the LoggerSilence to work as described:

- Following the Rails guide which state that a logger needs to include
  the `ActiveSupport::LoggerSilence` as well as
  `ActiveSupport::LoggerThreadSafe` modules isn't enough and won't
  work.

  Here is a test cases with 3 tests that all fails
  https://gist.github.com/Edouard-chin/4a72930c2b1eafbbd72a80c66f102010

  The problems are the following:

  1) The logger needs to call `after_initialize` in order to setup
  some instance variables.
  2) The silence doesn't actually work because the bare ruby Logger
  `add` method checks for the instance variable `@logger`. We need to
  override the `add` (like we used to in the ActiveSupport::Logger
  class).
  3) Calling `debug?` `info?` etc... doesn't work as the bare ruby
  methods will check for the instance variable. Again we need to
  override this methods (like we used to in the ActiveSupport::Logger
  class)

  The LoggerSilence won't work without LoggerThreadSafe, but the later
  is not public API, the user shouldn't have to include it so I
  modified to include it automatically.
  Same for the `after_initialize` method. I find unuintitive to have
  to call it directly. I modified to instance the variables when the
  module get included.

amatsuda committed:
:warning: assigned but unused variable - logger

amatsuda committed:
Bring config_accessor's API document back to its life

the document has been disappeared since the method became private at c2bfe6cbc8cab9caeab418472a1e12a3ed3e75e2

prathamesh-sonpatki committed:
Fix the docs for Module#parents and related methods in guides [ci skip]

- Followup of https://github.com/rails/rails/pull/34051

gmcgibbon committed:
Move test_fixtures and render_context to separate files

eileencodes committed:
Merge pull request #34061 from gmcgibbon/fixtures_refactor_file

Move test_fixtures and render_context to separate files

bogdanvlviv committed:
Fix rubocop issue

Fixes: `Layout/TrailingWhitespace: Trailing whitespace detected.

See https://codeclimate.com/github/rails/rails/issues

Releted to b707a6d0eb7

rafaelfranca committed:
Merge pull request #34063 from bogdanvlviv/fix-rubocop-offense-b707a6d0eb7

Fix rubocop issue

rafaelfranca committed:
Merge pull request #34055 from Edouard-chin/ec-logger-fix

Fix the LoggerSilence to work as described:

schneems committed:
Revert "Merge pull request #33970 from rails/eager-url-helpers"

Until #34050 can be resolved

This reverts commit 7f870a5ba2aa9177aa4a0e03a9d027928ba60e49, reversing
changes made to 6556898884d636c59baae008e42783b8d3e16440.

y-yagi committed:
Don't use deprecated `LoggerSilence`

schneems committed:
Merge pull request #33936 from schneems/schneems/cache-micro-optimizations

Decrease memory allocations in cache.rb

eugeneius committed:
Remove unreachable database warning

`establish_connection` will never raise `ActiveRecord::NoDatabaseError`,
because it doesn't connect to a database; it sets up a connection pool.

kamipo committed:
Merge pull request #34066 from eugeneius/remove_database_warning

Remove unreachable database warning

schneems committed:
This PR speeds up Nil#try by avoiding an allocation when only one argument is passed:

```ruby
class FooNew
  def try(method_name = nil, *args)
    nil
  end
end

class FooOld
  def try(*args)
    nil
  end
end

require 'benchmark/ips'


foo_new = FooNew.new
foo_old = FooOld.new

Benchmark.ips do |x|
  x.report("new") { foo_new.try(:anything) }
  x.report("old") { foo_old.try(:anything) }
  x.compare!
end
# Warming up --------------------------------------
#                  new   250.633k i/100ms
#                  old   232.322k i/100ms
# Calculating -------------------------------------
#                  new      6.476M (± 4.8%) i/s -     32.332M in   5.005777s
#                  old      5.258M (± 3.2%) i/s -     26.485M in   5.042589s

# Comparison:
#                  new:  6476002.5 i/s
#                  old:  5257912.5 i/s - 1.23x  slower
```

It's worth noting that checking for nil separately as in https://github.com/rails/rails/pull/34067 seems to be MUCH faster. It might be worth it to apply a blanket `&.` to every internal `try` call.

y-yagi committed:
Add `skip-webpack-install` option

This option is useful when want to check only the files generated by
`rails new`, or if want to do something before `webpacker:install`.

y-yagi committed:
Avoid `webpacker:install` if unnecessary

`webpacker:install` also includes execution of yarn, it takes time to execute,
so avoid unnecessary tests.

aidanharan committed:
Escape table name so that it can be used in regular expression

kamipo committed:
Merge pull request #34073 from aidanharan/activerecord-regex-escape-table-name

Escape table name so that it can be used in regular expression

eileencodes committed:
Merge pull request #33938 from faucct/bugfix/preload_through_no_records

ActiveRecord::Associations::Preloader should not fail to preload through missing records

hakusaro committed:
Change contributing guide to use safer force push

This commit changes the contributing guide in rails to suggest using git
with force-with-lease over typical force pushing. In practice, most
rails contributors won't ever encounter a situation where updating their
local fork could result in lost changes as a result of a force push.

That being said, git is a complex tool and arcane flags like
force-with-lease are indeed safer, and by promoting it in rails, there's
a chance more people will discover it and use it in other contexts
outside of rails. In just the same way that herd immunity works by most
people being vaccinated, proliferating knowledge of force-with-lease
should help nudge people towards using safer git commands in general.

[ci skip]

kamipo committed:
Restore `preloaders_for_one` method

Since the above comment is for the `preloaders_for_one`.

schneems committed:
Merge pull request #34064 from schneems/schneems/revert-7f870a5ba2aa9177aa4a0e03a9d027928ba60e49

Revert "Merge pull request #33970 from rails/eager-url-helpers"

kamipo committed:
Simplify the condition in `prepare_update_statement`

kamipo committed:
Use `assert_no_queries` not to ignore BEGIN/COMMIT queries

`test_update_does_not_run_sql_if_record_has_not_changed` would pass
without #18501 since `assert_queries` ignores BEGIN/COMMIT unless
`ignore_none: true` is given.

Since #32647, empty BEGIN/COMMIT is ommited. So we no longer need to use
`assert_queries(0)` to ignore BEGIN/COMMIT in the queries.

cllns committed:
Add instruction for test environment

kamipo committed:
Remove `ignore_none: false` to assert no queries more strictly

Follow up 811be477786455d144819a5e9fbb7f9f54b8da69.

gmcgibbon committed:
Add multi-db support to rails db:migrate:status

gmcgibbon committed:
Move db:migrate:status to DatabaseTasks method

bogdanvlviv committed:
Wrap custom id of a header attribute into \" in rails guides

Related to https://github.com/rails/rails/pull/33953

kamipo committed:
Merge pull request #34083 from bogdanvlviv/follow-up-33953

Wrap custom id of a header attribute into \" in rails guides

spickermann committed:
Bugfix: ActiveSupport::EncryptedConfiguration reading of comment-only encrypted files (#34014)

* Fix reading comment only encrypted files

When a encrypted file contains only comments then reading that files raises an error:

    NoMethodError: undefined method `deep_symbolize_keys' for false:FalseClass
        activesupport/lib/active_support/encrypted_configuration.rb:33:in `config'
        test/encrypted_configuration_test.rb:52:in `block in <class:EncryptedConfigurationTest>'

This happens because the previous implementation returned a `{}` fallback for blank YAML strings. But it did not handle YAML strings that are present but still do not contain any _usefull_ YAML - like the file created by `Rails::Generators::EncryptedFileGenerator` which looks like this:

    # aws:
    #   access_key_id: 123
    #   secret_access_key: 345

* Fix coding style violation

* Add backwardscompatible with Psych versions that were shipped with Ruby <2.5

* Do not rely on railties for Active Support test

* Simplify error handling

* Improve test naming

* Simplify file creation in test

y-yagi committed:
Avoid `webpacker:install` if unnecessary

Since this is a test to check the behavior of `load_defaults`,
webpacker is unnecessary.

y-yagi committed:
Don't use deprecated `Module#parents`

frodsan committed:
Add retry_on/discard suggestions for common cases.

This adds the same suggestions added here:
https://github.com/rails/rails/blob/01a69e27a4e55504af8fe776826d659550e6f89e/activejob/lib/rails/generators/job/templates/application_job.rb

These appear when `app/jobs/application_job.rb` doesn't exist, but not
for new applications.

gregmolnar committed:
update nokogiri

lsylvester committed:
add test for cache_version precision

y-yagi committed:
Merge pull request #34087 from frodsan/application_job_suggestions

Add retry_on/discard suggestions for common cases to ApplicationJob for new apps.

bogdanvlviv committed:
Exercise stringify of database configurations

Since #33968 we stringify keys of database configuration
This commit adds more assertions in order to ensure that and prevent any
regression in the future.
Currently, if remove `to_s` added in #33968 from `env_name.to_s` on the
line
(activerecord/lib/active_record/database_configurations.rb:107), there is
no test that would fail. One of the added assertions should emphasize why we need
this `to_s`.

Follow up #33968

chrismaximin committed:
Clear QueryCache when reloading associations

kamipo committed:
Merge pull request #34092 from bogdanvlviv/follow-up-33968

Exercise stringify of database configurations

schneems committed:
Merge pull request #34068 from schneems/schneems/micro-optimize-try-nil

23% faster Nil#try

prathamesh-sonpatki committed:
Fix tests related to new JavaScript path for generators

- Followup of https://github.com/rails/rails/commit/4838c1716a0340137d858fab49bf460e23be5a4b

guilleiguaran committed:
Merge pull request #34095 from prathamesh-sonpatki/fix-tests-related-to-ac

Fix tests related to new JavaScript path for generators

prathamesh-sonpatki committed:
Skip node_modules dir in the rubocop check

- Otherwise it is running the check against all files in node_modules

kule committed:
Parameterized mailers can configure delivery job

Setting parameterized_delivery_job on a mailer class will cause Parameterized::MessageDelivery to use
the specified job instead of ActionMailer::Parameterized::DeliveryJob:

    class MyMailer < ApplicationMailer
      self.parameterized_delivery_job = MyCustomDeliveryJob
      ...
    end

rafaelfranca committed:
Merge pull request #34096 from prathamesh-sonpatki/exclude-node-modules-for-rubocop

Skip node_modules dir in the rubocop check

gmcgibbon committed:
Remove bundler warnings from bin/setup output

gmcgibbon committed:
Organize FixtureSet class methods

gmcgibbon committed:
Move FixtureSet.create_fixtures reading and inserting sections into private methods

gmcgibbon committed:
Small refactors to FixtureSet::ClassCache and Fixture

gmcgibbon committed:
Introduce FixtureSet::ModelMetadata

gmcgibbon committed:
Introduce FixtureSet::TableRows and FixtureSet::TableRow

gmcgibbon committed:
Move FixtureSet::ReflectionProxy and FixtureSet::HasManyThroughProxy to FixtureSet::TableRows

gmcgibbon committed:
Remove some autolabeling globs

[ci skip]

rafaelfranca committed:
Merge pull request #34102 from gmcgibbon/less_aggressive_auto_labeling

Remove some autolabeling globs

albertoalmagro committed:
Add documentation to run tests with specific seed [ci skip]

This commit documents how to run all tests or a single test file with a
specific randomization seed.

rafaelfranca committed:
Merge pull request #34097 from kule/master

Parameterized mailers can configure delivery job

rafaelfranca committed:
Merge pull request #34101 from gmcgibbon/fix_bundler_1_16_6_test_failures

Remove bundler warnings from bin/setup output

XrXr committed:
Remove unnecessary use of `included` in ActiveJob::Core

Using `included` to define `attr_acessor` and `attr_writer` is causing
these methods to not show up in the documentation.

zvkemp committed:
clarify role of unique_id in ActiveSupport::Notifications

gmcgibbon committed:
Merge pull request #34103 from albertoalmagro/guides-running-test-with-seed

Add documentation to run tests with specific seed [ci skip]

rafaelfranca committed:
Merge pull request #34084 from trustvox/master

Remove "include FileUtils" sentence in setup/update bin files

schneems committed:
Merge pull request #34005 from tgturner/expand-cache-key-bug-fix

Array with single item correctly uses cache_key

y-yagi committed:
Merge pull request #34077 from cllns/clarify-activesupport-instruction

ActiveStorage guide: Add instruction for test environment

bogdanvlviv committed:
Follow up #34064

The removed code was added in 7f870a5ba2aa9177aa4a0e03a9d027928ba60e49,
then 7f870a5ba2aa9177aa4a0e03a9d027928ba60e49 was reverted by #34064.
But I found that that commit wasn't completely reverted, I guess it was
caused by resolving some conflicts during reverting.
@schneems could you please confirm that those changes weren't reverted
unintentionally, or reject this commit otherwise?

albertoalmagro committed:
Raise on invalid definition values

When defining a Hash enum it can be easy to use [] instead of {}. This
commit checks that only valid definition values are provided, those can
be a Hash, an array of Symbols or an array of Strings. Otherwise it
raises an ArgumentError.

Fixes #33961

albertoalmagro committed:
Privatize ENUM_CONFLICT_MESSAGE constant

schneems committed:
Merge pull request #34089 from lsylvester/test-cache-version-precision

add test for cache_version precision

madogiwa0124 committed:
getting started page rails version update to 5.2.1

kamipo committed:
Fix `AssociationRelation` not to set inverse instance key just like before

Since #31575, `set_inverse_instance` replaces the foreign key by the
current owner immediately to make it happen when a record is added to
collection association.

But `set_inverse_instance` is not only called when a record is added,
but also when a record is loaded from queries. And also, that loaded
records are not always associated records for some reason (using `or`,
`unscope`, `rewhere`, etc).

It is hard to distinguish whether or not we should invoke
`set_inverse_instance`, but at least we should avoid the undesired
side-effect which was brought from #31575.

Fixes #34108.

eileencodes committed:
Merge pull request #34114 from Madogiwa0124/update_getting_start_rails_version

getting started page rails version update to 5.2.1

eileencodes committed:
Merge pull request #34076 from gmcgibbon/fixtures_refactor

Fixtures refactor

kamipo committed:
Fix test name to add missing "set"

conzett committed:
Fix issue ActiveStorage direct upload disk

Fix an issue in ActiveStorage where a direct upload to disk storage
would fail due to a content type mismatch if the file was uploaded using
a mime-type synonym.

aergonaut committed:
Add docs to ActiveRecord::Persistence#belongs_to_touch_method

[ci skip]

frodsan committed:
[ci skip] Fix typo

arunagw committed:
Merge pull request #34119 from frodsan/patch-1

[ci skip] Fix typo

kamipo committed:
Generate delegation methods to named scope in the definition time

The delegation methods to named scope are defined when `method_missing`
is invoked on the relation.

Since #29301, the receiver in the named scope is changed to the relation
like others (e.g. `default_scope`, etc) for consistency.

Most named scopes would be delegated from relation by `method_missing`,
since we don't allow scopes to be defined which conflict with instance
methods on `Relation` (#31179). But if a named scope is defined with the
same name as any method on the `superclass` (e.g. `Kernel.open`), the
`method_missing` on the relation is not invoked.

To address the issue, make the delegation methods to named scope is
generated in the definition time.

Fixes #34098.

kamipo committed:
Merge pull request #34039 from yskkin/parameter_filter

Deprecate ActionDispatch::Http::ParameterFilter in favor of ActiveSup…

kamipo committed:
Merge pull request #33324 from Jcambass/fix-only-path-option-in-url-for-with-arrays

respect only_path option when an array is passed into url_for

conzett committed:
Use content_mime_type

gregmolnar committed:
fix broken link in Action Cable guides and readme [ci skip]

georgeclaghorn committed:
Fix directly uploading using a MIME type synonym

When Content-Type is "application/x-gzip", request.content_type resolves to "application/gzip", because application/x-gzip is a synonym of application/gzip by default. This causes the acceptable_content? check in ActiveStorage::DiskController to fail, because the direct upload token contains application/x-gzip, which is not equal to application/gzip.

Fix by comparing the token content type with the request content type *and its synonyms*.

prodion23 committed:
Point to requiring the ASt engine in the installation instructions [ci skip]

kamipo committed:
Merge pull request #34126 from gregmolnar/guide-link-fix

fix broken link in Action Cable guides and readme [ci skip]

kamipo committed:
Call `define_attribute_methods` before `assert_no_queries` to address CI flakiness

Follow up 45be690f8e6db019aac6198ba49d608a2e14824b.

Somehow calling `define_attribute_methods` in `build`/`new` sometimes
causes the `table_exists?` query.

To address CI flakiness due to `assert_no_queries` failure, ensure
`define_attribute_methods` before `assert_no_queries`.

y-yagi committed:
Use `--skip-webpack-install` by default

To remove extra `--no-skip-javascript` tests.

kamipo committed:
Merge pull request #34117 from aergonaut/docs/ActiveRecord--Persistence-belongs_to_touch_method

Add docs to ActiveRecord::Persistence#belongs_to_touch_method

[ci skip]

gmcgibbon committed:
Merge pull request #34105 from zvkemp/correct-asn-docs

clarify role of unique_id in ActiveSupport::Notifications [ci skip]

eileencodes committed:
Add allocations to template renderer subscription

This PR adds the allocations to the instrumentation for template and
partial rendering.

Before:

```
  Rendering posts/new.html.erb within layouts/application
  Rendered posts/_form.html.erb (9.7ms)
  Rendered posts/new.html.erb within layouts/application (10.9ms)
Completed 200 OK in 902ms (Views: 890.8ms | ActiveRecord: 0.8ms)
```

After:

```
  Rendering posts/new.html.erb within layouts/application
  Rendered posts/_form.html.erb (Duration: 7.1ms | Allocations: 6004)
  Rendered posts/new.html.erb within layouts/application (Duration: 8.3ms | Allocations: 6654)
Completed 200 OK in 858ms (Views: 848.4ms | ActiveRecord: 0.4ms | Allocations: 1539564)
```

eileencodes committed:
Merge pull request #34081 from gmcgibbon/db_migrate_status_move

Move db:migrate:status to DatabaseTasks method

gmcgibbon committed:
Merge pull request #34075 from hakusaro/guides-reference-leases

Change contributing guide to suggest using safer force push

eileencodes committed:
Merge pull request #34071 from y-yagi/skip_webpacker_install

Add `skip-webpack-install` option

kamipo committed:
Improve DELETE with JOIN handling to avoid subqueries if possible

Before:

```
  Pet Destroy (0.8ms)  DELETE FROM `pets` WHERE `pets`.`pet_id` IN (SELECT `pet_id` FROM (SELECT DISTINCT `pets`.`pet_id` FROM `pets` LEFT OUTER JOIN `toys` ON `toys`.`pet_id` = `pets`.`pet_id` WHERE `toys`.`name` = ?) AS __active_record_temp)  [["name", "Bone"]]
```

After:

```
  Pet Destroy (1.0ms)  DELETE `pets` FROM `pets` LEFT OUTER JOIN `toys` ON `toys`.`pet_id` = `pets`.`pet_id` WHERE `toys`.`name` = ?  [["name", "Bone"]]
```

kamipo committed:
Refactor Arel visitor to use `collect_nodes_for` as much as possible

kamipo committed:
Fix odd indentation

kamipo committed:
Call `load_schema` before `assert_no_queries`

Follow up 45be690f8e6db019aac6198ba49d608a2e14824b.

`predicate_builder.build` in `where` requires `load_schema` for
`type_for_attribute`.

kamipo committed:
Merge pull request #34094 from christophemaximin/fix-activerecord-clearing-of-query-cache

Fix inconsistent behavior by clearing QueryCache when reloading associations

kamipo committed:
Merge pull request #34122 from kamipo/generate_relation_methods

Generate delegation methods to named scope in the definition time

eileencodes committed:
Merge pull request #34136 from rails/add-allocations-to-template-renderer

Add allocations to template renderer subscription

eileencodes committed:
Merge pull request #34109 from bogdanvlviv/follow-up-34064

Follow up #34064

eileencodes committed:
Merge pull request #34137 from gmcgibbon/db_migrate_status_multi_db

Add multi-db support to rails db:migrate:status

eileencodes committed:
Merge pull request #34110 from albertoalmagro/enum-raises-on-invalid-definition-values

Enum raises on invalid definition values

eileencodes committed:
Merge pull request #34052 from eileencodes/connection-switching

Part 4: Multi db improvements, Basic API for connection switching

Commit:
Remove invalid magic comment [ci skip]

kamipo committed:
Merge pull request #34188 from yskkin/remove_comment

Remove invalid magic comment [ci skip]

orhantoy committed:
[ci skip] Fix link to Concurrent::ThreadPoolExecutor docs

kaspth committed:
Merge pull request #34189 from orhantoy/fix-link-to-concurrent-ruby-docs

[ci skip] Fix link to Concurrent::ThreadPoolExecutor docs

multiplegeorges committed:
Merge changelog conflict.

multiplegeorges committed:
Fix conflict.
